code/game/mecha/mecha_actions.dm

```diff
@@ -164,10 +164,31 @@
 		chassis.log_message("Toggled thrusters.", LOG_MECHA)
 		chassis.occupant_message("<font color='[chassis.thrusters_active ?"blue":"red"]'>Thrusters [chassis.thrusters_active ?"en":"dis"]abled.")
 
+/**An object to take the hit for us when using the Durand's defense mode.
+It is spawned by the Durand (see the durand.dm file) when an attack is made and determined to be blocked by the shield.
+Attacks are passed to it after creation. It will deal the damage as power drain to the mech and then self-delete.*/
+/obj/mech_defence_mode_scapegoat //projectiles get passed to this when defence mode is enabled
+	name = "defence grid"
+	max_integrity = 10000
+	obj_integrity = 10000
+	var/obj/mecha/chassis
+
+
+/obj/mech_defence_mode_scapegoat/take_damage()
+	if(!chassis)
+		qdel(src)
+		return
+	. = ..()
+	if(!chassis.use_power((max_integrity - obj_integrity) * 100))
```

Comment:
Less damage drains more power? Something is off with this calculation.

Reply:
If I have 20 max health, and I lose 5, my health is left at 15. So 20 - 15 is 5 charge to lose.
If I have 20 max health, and I lose 7, my health is left at 14. So 20 - 14 is 7 charge to lose.
We then multiply it by 100 to get the actual charge used.

If chassis.use_power returns false, it's because the amount of power needed surpassed what was available. The cell proc that handles using power will just output a false in this situation, but doesn't drain the power at all. So instead we have to do it manually.

Reply:
Yeah sorry, I forgot what a obj_integrity did. Not like this proc makes sense anyways. Should probably just be (. * 100) though.

Reply:
what if i had 20 max health 10 health, lose 1, do i lose 1 power or 11?

Reply:
The way this works is that a new entity is spawned in (with a shit load of health), damage is dealt to it, and then after calculating the charge to take, the entity is removed. The next time damage is done, a new entity is spawned in with a fresh health bar, to act as the scapegoat.

It's honestly probably not the sanest way to go about it, and I might change it. It was just a decent way in my mind of passing on damage and calculating from it without having to worry about how different damage types might affect things.

Reply:
You can basically abstract this by moving this code to take_damage.
Something like this:
```
/obj/mecha/combat/durand/take_damage(damage_amount, damage_type = BRUTE, damage_flag = "", sound_effect = TRUE, attack_dir, armour_penetration = 0)
if(defence_check(damage_amount, attack_dir, damage_flag))
	return
. = ..()

/obj/mecha/combat/durand/proc/defence_check(damage, adir, damage_flag = "")
	if (!defence_mode)
		return
	if(damage_flag == "bomb")
		return
	switch(dir) //There's probably a better way, haven't checked.
		if (EAST)
			if(adir != WEST)
				return
		if (SOUTH)
			if(adir != NORTH)
				return
		if (WEST)
			if(adir != EAST)
				return
		if (NORTH)
			if(adir != SOUTH)
				return
	if(!use_power(damage * 100))
		defense_action.turnoff()
		return
	return TRUE

Reply:
if(damage_flag == "stamina")
		return

Reply:
That's nonsense. damage_flag is basically the type of armor you want to use. And objs already don't get damaged by anything that isn't BRUTE or BURN.

Reply:
After toying with it for a bit, I'm not sure I want to re-do it like this.

The biggest issue is that diagonals don't neatly fall into a dir, so they're sorta fudged a bit. When you fire in a diagonal, the projectile is given a four-way-compatible dir, usually confined to one axis (I believe left and right). The result is that if the mech is facing two specific directions (I believe north or south), its field of defense will be a lot narrower than if it was facing the other directions. Ideally, I'd want the mech to perform consistently no matter what direction its facing.

That aside, melee attacks don't seem to give a dir, so I'd have to add extra checks in for that.